Windows CI sometimes hangs

This appears to only happen when the net tests are being run (by default they are off for now).

While testing windows networking in general, I tried putting all the TCP tests on the same port (60000) and the workflow hung on exit. Note this is with the tests run sequentially.  All of the tests ran to completion but until I cancelled, the output was stuck.

```
2023-09-02T03:09:05.6454013Z    Creating library D:\a\ponyc\ponyc\build\debug\stdlib-release.lib and object D:\a\ponyc\ponyc\build\debug\stdlib-release.exp
2023-09-02T03:09:07.6887440Z C:\msys64\mingw64\bin\lldb.exe --batch --one-line run --one-line-on-crash "frame variable" --one-line-on-crash "register read" --one-line-on-crash "bt all" --one-line-on-crash "quit 1" -- D:\a\ponyc\ponyc\build\debug\stdlib-release.exe --sequential --only="net/"
2023-09-02T03:34:10.0822541Z (lldb) target create "D:\\a\\ponyc\\ponyc\\build\\debug\\stdlib-release.exe"
```

It almost appears as if everything works BUT, somehow we are getting something for the workflow hung.

The end was...

```
2023-09-02T03:34:10.0907443Z [91m**** FAILED: net/TCP.writev[0m
2023-09-02T03:34:10.0907825Z Process 1880 exited with status = -1 (0xffffffff)
2023-09-02T03:34:10.0908348Z Process 1880 launched: 'D:\a\ponyc\ponyc\build\debug\stdlib-release.exe' (x86_64)
2023-09-02T03:34:10.4005012Z grammar original hash:  E514F31713CA2D2DE95FD4CFB889B83B4063AB2F2BC931DB90BCD67D63678F0F
2023-09-02T03:34:10.4006763Z grammar generated hash: E514F31713CA2D2DE95FD4CFB889B83B4063AB2F2BC931DB90BCD67D63678F0F
2023-09-02T03:34:10.4042752Z Test suites run: 7, num failed: 2
2023-09-02T03:34:10.4088086Z Test suites failed: (stdlib-debug, stdlib-release)
2023-09-02T03:34:10.4106853Z 
2023-09-02T03:34:10.4420234Z ##[error]Process completed with exit code 1.
2023-09-02T03:34:10.4700821Z Post job cleanup.
``` 

So... it did run to completion, it just "got stuck outputting" until cancelled. Very odd, but it does look as though running the net tests all on the same port makes this predictable. However, I have other change related to net test flakiness that are part of the branch I'm on so perhaps not.

Related to #4413

--- a/.release-notes/4437.md
+++ b/.release-notes/4437.md
@@ -1,0 +1,15 @@
+## Fix "double socket close" issue with Windows version of TCPConnection
+
+When fixing a number of "smaller" Windows TCP networking issues a couple years back, in addition to fixing those issues, we introduced a new bug. That bug lingered for two years. It lingered in large part because it would only become apparent in a low resource environment.
+
+When we recently switched our Windows CI from CirrusCI to GitHub Actions, we went from a high-resource environment to a low-resource environment and started getting a ton of "random" Windows TCP test failures.
+
+The problem that was fairly easy to recreate in a test environment would be fairly unlikely in most applications but existed nontheless. The scenario in our test environment was like this:
+
+- Test 1 runs and completes but hasn't done test teardown yet
+- Test 2 starts up
+- Test 1 runs the "buggy" line of code and closes the socket it has been using with OS, but doesn't reset its own internal record of the file descriptor for the socket.
+- Test 2 is gets a socket from the OS with the file descriptor for the socket just closed by Test 1
+- Test 1 still has a "valid" file descriptor and as part of full shutdown, closes the socket associated with "its" file descriptor. When Test 1 does this, test 2's socket closes and the test fails to complete successfully.
+
+The problem would appear "in the wild" if a Windows application was quickly closing and opening TCP sockets in a manner similiar to the Pony standard library TCP tests.

--- a/packages/net/tcp_connection.pony
+++ b/packages/net/tcp_connection.pony
@@ -1059,11 +1059,7 @@ actor TCPConnection is AsioEventNotify
       _shutdown = true
 
       if _connected then
-        ifdef windows then
-          @pony_os_socket_close(_fd)
-        else
-          @pony_os_socket_shutdown(_fd)
-        end
+        @pony_os_socket_shutdown(_fd)
       else
         _shutdown_peer = true
       end

